### DISCUSSION TITLE
Multiple Instances for the Same Animal, Even Though Max Instances is Restricted in Training Pipeline

### DISCUSSION THREAD

**Discussion by majaneubauer** [talmolab/sleap#1676] [Help!] [answered]
I am currently using Google Colab and SLEAP (Version 1.3.3) to train a bottom-up multi-animal model. I have videos of two mice (one black and one white) in an open field arena, filmed from below. I am training my model so that it is able to identify the two animals in my video. So far I have trained on 515 frames. However, I have run into some issues:

When predicting on a new video after training, I often get up to four predicted instances for the same animal (see attached screenshot). What I have been doing so far is double-clicking one of the instances and correcting the predictions, while ignoring the other instances (see screenshot). I assumed that the yellow "not in training data" means that these predictions won't be used in future training.
![image](https://github.com/talmolab/sleap/assets/131380612/1d029903-18a1-4fe1-ae8f-690c246a391a)
![image](https://github.com/talmolab/sleap/assets/131380612/760b1be2-dfa5-4ef6-afb7-cb9739a88a86)

After looking into this issue I realised that I can restrict the max instances when exporting my training package. I set it to 2 and predicted again on my user labelled frames (see Training pipeline screenshot). However, I still get up to four predicted instances for the same animal.
![image](https://github.com/talmolab/sleap/assets/131380612/0fb169bc-1426-4293-b344-e8a959052d60)

I now have three questions:

1. Is there another way to restrict how many animals to get predicitions on, i.e. tell the model that there are only two mice in my videos?

2. Am I correct in my assumption that the other ignored instances (labelled "not in training data") will in fact not get used for future training? 

3. Do you have any suggestions as to what I should be doing to make my model realise that there are only two animals, and therefore there should only be two instances per frame?

Thank you!

**Comment by roomrys** [comment]
Hi @majaneubauer,

Update
---
The user was running tracking through the CLI! There was a missing argument to specify which type of tracker to use  (`flowmaxtracks`) which uses/implements the max tracking. See designated [answer](https://github.com/talmolab/sleap/discussions/1676#discussioncomment-8399576) from user.

---

After looking at your discussion -  we did a little test case over here. Similar to you, I trained a bottom-up (non-supervised-id) model on a dataset with two animals. I then ran inference on "user-labeled frames" two different times. The first time, I ran inference with max instances set to "No max" and 2 instances were predicted on my frame - as expected. The second time, I ran inference on the same "user-labeled frames" with max instances set to 1 and only 1 instance was predicted on my frame - also as expected. But not great for recreating the issue or figuring out what is goin on.

## Max of 1 predicted instance (yellow-boxed instance) per user-labeled frame
I am fairly sure that I checked that max instances was working before releasing... but perhaps we did not identify an edge case that you are running into.
![image](https://github.com/talmolab/sleap/assets/38435167/9dfcd1a5-d716-4eeb-b74b-a9e5f65410c4)
![image](https://github.com/talmolab/sleap/assets/38435167/146d7dc5-e20e-42a5-82ab-28f7b280ac91)
![image](https://github.com/talmolab/sleap/assets/38435167/6fa92921-1095-4d7f-83fd-473b6529eea1)

## Inference configuration
I call inference as part of the Inference Pipeline (which uses an already trained model) whereas you have run inference as the next step after training a new model in the Training Pipeline - I should do that to see if anything is amiss there.
![image](https://github.com/talmolab/sleap/assets/38435167/6ff68e2f-8375-40ba-8249-e9d5c73c9ee7)

## After running inference another time on the same frames, we see that the previous predictions were replaced with new predictions
I though maybe there were predictions left over from another inference run and that might be why there were more predictions than we had specified. Although I was pretty sure we remove predictions before adding new ones, I double checked the code to verify and ran a test which show that: yes, predictions are indeed removed before adding new predicitons.
![image](https://github.com/talmolab/sleap/assets/38435167/d73463a4-dc81-407c-9d01-1445d174d1b4)


A few differences between our runs
1. You are running inference as part of the training process whereas I ran inference on it's own
2. You are on the official 1.3.3 release, I am on develop (but we don't seem to have any track related [commits since 1.3.3](https://github.com/talmolab/sleap/commits/develop/?since=2023-09-15&until=2024-02-07) so this should not be the cause)

~~I will now re-run training + inference to see if there is a difference and update below. In the meantime, I would encourage running just inference (not as part of the training) with the max instances set.~~

~~... updates still loading ...~~

Thanks,
Liezl

**Reply by majaneubauer** [reply, marked as answer]
Hi @roomrys,

Thank you for your quick reply! 

I have been doing some testing on my own the past two days and just now realised where I went wrong. Exporting the training package specifying the max instances to 2 and predicting on user-labeled frames was not the problem. Rather, I made a mistake while running inference on Colab using `sleap-track`. I reported my issue in [#1679](https://github.com/talmolab/sleap/discussions/1679), which I have now closed. I had to specify my `tracking.tracker` to `flowmaxtracks` in order for `tracking.max_tracking` to work. 

I will go ahead and close this discussion as well now, but thank you for your efforts, I appreciate it!

Best,
Maja

**Reply by roomrys** [reply]
Hi @majaneubauer,

Awesome news, thanks for following up on that. The max tracks is a somewhat new feature which I have not had much interaction with yet - totally forgot about the `flowmaxtracks`! But, I will be remembering/double-checking that from now on.

Not so quick, but always trying to be better - appreciate the positive reinforcement!

Thanks,
Liezl
